subcommand/acl-controller/command.go

```diff
@@ -185,6 +295,131 @@ var partitionedClientPolicyTpl = `partition "%s" {
+	}, c.writeOptions())
+	if err != nil {
+		return fmt.Errorf("creating Consul client ACL policy: %w", err)
+	}
+	c.log.Info("ACL policy created successfully", "name", policyName)
+	return nil
+}
+
+// upsertClientRole creates the ACL role for the Consul client, if the role does not exist.
+func (c *Command) upsertClientRole(consulClient *api.Client, roleName, policyName string) error {
+	// If the role already exists, we're done.
+	role, _, err := consulClient.ACL().RoleReadByName(roleName, c.queryOptions())
+	if err != nil && !controller.IsACLNotFoundError(err) {
+		return fmt.Errorf("reading Consul client ACL role: %w", err)
+	} else if err == nil && role != nil { // returns role=nil and err=nil if not found
+		c.log.Info("ACL role already exists; skipping role creation", "name", roleName)
+
+		if len(role.Policies) == 0 {
+			c.log.Info("updating ACL role with policy", "role", roleName, "policy", policyName)
+			role.Policies = []*api.ACLLink{{Name: policyName}}
+			_, _, err := consulClient.ACL().RoleUpdate(role, c.writeOptions())
+			if err != nil {
+				return fmt.Errorf("updating Consul client ACL role: %s", err)
+			}
+			c.log.Info("update ACL role successfully", "name", roleName)
+		}
+		return nil
+	}
+
+	c.log.Info("creating ACL role", "name", roleName)
+	_, _, err = consulClient.ACL().RoleCreate(&api.ACLRole{
+		Name:        roleName,
+		Description: "Consul Client Token Role for ECS",
+		Policies: []*api.ACLLink{
+			{
+				Name: policyName,
+			},
+		},
+	}, c.writeOptions())
+	if err != nil {
+		return fmt.Errorf("creating Consul client ACL role: %w", err)
+	}
+	c.log.Info("ACL role created successfully", "name", roleName)
+
+	return nil
+}
+
+func (c *Command) upsertAuthMethod(consulClient *api.Client, authMethod *api.ACLAuthMethod) error {
+
+	method, _, err := consulClient.ACL().AuthMethodRead(authMethod.Name, c.queryOptions())
+	if err != nil && !controller.IsACLNotFoundError(err) {
+		return fmt.Errorf("reading ACL auth method: %w", err)
+	} else if err == nil && method != nil { // returns err=nil and method=nil if not found
+		c.log.Info("ACL auth method already exists; skipping creation", "name", authMethod.Name)
+		return nil
+	}
+
+	c.log.Info("creating ACL auth method", "name", authMethod.Name)
+	method, _, err = consulClient.ACL().AuthMethodCreate(authMethod, c.writeOptions())
+	if err != nil {
+		return fmt.Errorf("creating ACL auth method: %w", err)
+	}
+	c.log.Info("ACL auth method created successfully", "name", method.Name)
+
+	return nil
+}
+
+func (c *Command) upsertBindingRule(consulClient *api.Client, bindingRule *api.ACLBindingRule) error {
+	method := bindingRule.AuthMethod
+
+	rules, _, err := consulClient.ACL().BindingRuleList(method, c.queryOptions())
+	if err != nil {
+		return fmt.Errorf("listing ACL binding rules for auth method %s: %w", method, err)
+	}
+	if len(rules) > 0 {
+		// For now, we just expect at least one binding rule to exist.
+		// TODO: Can we create the rule with a client-generated ID?
```

Comment:
Based on glancing through the code, it looks like we can [set `ID` on the `ACLBindingRule`](https://github.com/hashicorp/consul/blob/main/api/acl.go#L222) and then call `BindingRuleRead` with that `ID` rather than using `BindingRuleList`. I think either approach is probably fine though.

Reply:
Yeah, I think this is a matter of which specific unique ID to use. I think `<account>-<region>-<cluster>` would work to uniqueify these resources (or a hash of that), and that would aslo avoid requiring an extra `-auth-method-prefix` flag.

I'll do that in a follow up though.